### PR TITLE
ISPN-11832 ProtobufMetadataManagerInterceptor is blocking

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-11832

### CODE REVIEW

**Comment by anistor** [comment]
Code looks good, makes sense, but there's a related failure I think : ProtobufMetadataManagerInterceptorTest.testStateTransfer

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryServerBlockHoundIntegration.java [hunk 1]
```diff
@@ -10,12 +9,5 @@
 public class RemoteQueryServerBlockHoundIntegration implements BlockHoundIntegration {
```
Comment (anistor) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryServerBlockHoundIntegration.java:9]: I'd remove this class entirely

**Comment by anistor** [comment]
Still waiting for CI :recycle:

**Comment by anistor** [comment]
@danberindei wdyt ?

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 2]
```diff
@@ -472,7 +523,7 @@ private void updateGlobalErrors(InvocationContext ctx, Set<String> errorFiles, l
          cmd = commandsFactory.buildPutKeyValueCommand(ERRORS_KEY_SUFFIX, sb.toString(),
                keyPartitioner.getSegment(ERRORS_KEY_SUFFIX), DEFAULT_METADATA, flagsBitSet);
       }
-      invoker.running().invoke(ctx, cmd);
+      return invoker.running().invokeAsync(ctx, cmd);
```
Comment (danberindei) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:526]: There is an older issue here (also present in the old blocking code).

The global errors key should contain the errors in all the schemas, not just the errors from the latest schema added/updated. So the interceptor can't use a `PutKeyValueCommand`, it should use a `ComputeCommand` or similar to remove the successful keys and add the new error keys to the global errors atomically.

Comment (anistor) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:526]: Old indeed. Now I remember it. I hope there's a JIRA for it.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryServerBlockHoundIntegration.java (hunk 1 above)
Comment (wburns) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryServerBlockHoundIntegration.java:9]: Okay.

**Comment by wburns** [comment]
So, it seems from comments that we can integrate this then?

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 3]
```diff
@@ -276,9 +285,32 @@ private void handlePutKeyValueResult(InvocationContext rCtx, PutKeyValueCommand
          }
 
          if (progressCallback != null) {
-            updateGlobalErrors(rCtx, progressCallback.getErrorFiles(), flagsBitSet);
+            CompletionStage<Void> schemaUpdate = updateSchema(rCtx, progressCallback);
```
Comment (danberindei) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:288]: IMO having a separate `if` here makes sense only if you move this line out of the `if` below:

```
registerProtoFile((String) key, (String) value, progressCallback);
```

I'd rather move the lines of this `if` block inside the `if` block above that assigns a non-null value to `progressCallback` instead.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 4]
```diff
@@ -276,9 +285,32 @@ private void handlePutKeyValueResult(InvocationContext rCtx, PutKeyValueCommand
...
+            CompletionStage<Object> errorUpdate = updateGlobalErrors(rCtx, progressCallback.getErrorFiles().keySet(), flagsBitSet);
+
+            return asyncValue(CompletionStages.allOf(schemaUpdate, errorUpdate))
+                  .thenApplyMakeStage(rCtx, putKeyValueCommand, (rCtx2, rCommand2, rv2) -> rv);
          }
       }
+      return SyncInvocationStage.makeStage(rv);
+   }
+
+   CompletionStage<Void> updateSchema(InvocationContext ctx, ProgressCallback progressCallback) {
```
Comment (danberindei) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:298]: I think `updateSchemaErrors` would be a bit better

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 5]
```diff
@@ -337,66 +369,81 @@ public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) {
...
+         return invokeNext(ctx, command);
+      }
+      if (!(command.getKey() instanceof String)) {
+         throw log.keyMustBeString(command.getKey().getClass());
+      }
+      String key = (String) command.getKey();
+      if (!shouldIntercept(key)) {
+         return invokeNext(ctx, command);
+      }
+      // lock .errors key
```
Comment (danberindei) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:393]: I suggest replacing all the instances of the "lock .errors key" comment with "lock global errors key", the fact that `.errors` is the key for global errors isn't obvious.

Maybe even extract a method `lockGlobalErrors` to be used everywhere.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 2 above)
Comment (danberindei) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:526]: I created https://issues.redhat.com/browse/ISPN-12631

**Review by danberindei — CHANGES_REQUESTED** [review]
Looks good @wburns , I only have a few minor comments.

I created a separate issue (https://issues.redhat.com/browse/ISPN-12631) for the global errors update.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 3 above)
Comment (wburns) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:288]: Sure. Also can remove the `else` block as well.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 5 above)
Comment (wburns) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:393]: This isn't really related to my changes, other than indentation. I can make the text change as I don't want to mess something up while I am gone.

**Comment by wburns** [comment]
Updated.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 6]
```diff
@@ -436,9 +479,13 @@ public Object visitReplaceCommand(InvocationContext ctx, ReplaceCommand command)
             }
 
             if (progressCallback != null) {
-               updateGlobalErrors(rCtx, progressCallback.getErrorFiles(), flagsBitSet);
+               CompletionStage<Void> schemaUpdate = updateSchemaErrors(rCtx, progressCallback);
```
Comment (danberindei) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:482]: You can move the lines inside the if above, just like in `handlePutKeyValueResult`

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 5 above)
Comment (danberindei) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java]: Yeah, it's not your comment, it was just annoying me every time I read it :)

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 6 above)
Comment (wburns) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:482]: Okay, it is moved. I don't want to move the one in `visitPutMapCommand` though as it has other code in between that I am not exactly familiar with and don't want to mess with the logic this close to release.

**Comment by danberindei** [comment]
Merged, thanks Will!

**Pull request merged**